hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/publishing/handlers/end/MessageErrorProcessor.java

```diff
@@ -13,39 +13,46 @@
 
 import javax.ws.rs.core.MediaType;
 import java.io.IOException;
-import java.nio.charset.Charset;
+import java.util.Map;
 import java.nio.charset.StandardCharsets;
 
 import static pl.allegro.tech.hermes.api.ErrorCode.INTERNAL_ERROR;
 import static pl.allegro.tech.hermes.api.ErrorDescription.error;
 import static pl.allegro.tech.hermes.common.http.MessageMetadataHeaders.MESSAGE_ID;
 import static pl.allegro.tech.hermes.frontend.publishing.handlers.end.RemoteHostReader.readHostAndPort;
+import static pl.allegro.tech.hermes.frontend.publishing.metadata.HeadersToMapTransformer.toHeadersMap;
 
 public class MessageErrorProcessor {
     private static final Logger logger = LoggerFactory.getLogger(MessageErrorProcessor.class);
     private final ObjectMapper objectMapper;
     private final Trackers trackers;
     private final HttpString messageIdHeader = new HttpString(MESSAGE_ID.getName());
+    private final TrackingHeadersExtractor trackingHeadersExtractor;
 
-    public MessageErrorProcessor(ObjectMapper objectMapper, Trackers trackers) {
+    public MessageErrorProcessor(ObjectMapper objectMapper, Trackers trackers,
+                                 TrackingHeadersExtractor trackingHeadersExtractor) {
         this.objectMapper = objectMapper;
         this.trackers = trackers;
+        this.trackingHeadersExtractor = trackingHeadersExtractor;
     }
 
     public void sendAndLog(HttpServerExchange exchange, Topic topic, String messageId, ErrorDescription error) {
         sendQuietly(exchange, error, messageId, topic.getQualifiedName());
-        log(error.getMessage(), topic, messageId, readHostAndPort(exchange));
+        log(error.getMessage(), topic, messageId, readHostAndPort(exchange),
+                trackingHeadersExtractor.extractHeadersToLog(toHeadersMap(exchange.getRequestHeaders())));
```

Comment:
What do you think about passing `HeaderMap` to `extractHeadersToLog` instead using `HeadersToMapTransformer.toHeadersMap` in every place where we call the `extractHeadersToLog` method?

Reply:
Yes, I thought about it but I decided to do similar interface to the `HeadersPropagator` which operates on pure `Map<String, String>`. Yes, it looks like too much of duplication. I'll change it to operate on request `exchange`